web/src/timerender.ts

```diff
@@ -24,6 +24,94 @@ export function clear_for_testing(): void {
+
+// Translates Zulip-specific format names, documented in the comments
+// below, into the appropriate options to pass to the Intl library
+// along with the user's locale to render date in that style of format.
+//
+// Note that because date/time formats vary with locale, the below
+// examples are what a user with English as their language will see
+// but users in other locales will see something different, especially
+// for any formats that display the name for a month/weekday, but
+// possibly in more subtle ways for languages with different
+// punctuation schemes for date and times.
+function get_format_options_for_type(type: DateOrTimeFormat): Intl.DateTimeFormatOptions {
+    const is_twenty_four_hour_time = user_settings.twenty_four_hour_time;
+
+    const time_format_options: Intl.DateTimeFormatOptions = is_twenty_four_hour_time
+        ? {hourCycle: "h23", hour: "2-digit", minute: "2-digit"}
+        : {
+              hourCycle: "h12",
+              hour: "numeric",
+              minute: "2-digit",
+          };
+
+    const weekday_format_options: Intl.DateTimeFormatOptions = {weekday: "long"};
+    const full_format_options: Intl.DateTimeFormatOptions = {dateStyle: "full"};
+
+    const dayofyear_format_options: Intl.DateTimeFormatOptions = {day: "numeric", month: "short"};
+    const dayofyear_year_format_options: Intl.DateTimeFormatOptions = {
+        ...dayofyear_format_options,
+        year: "numeric",
+    };
+    const long_format_options: Intl.DateTimeFormatOptions = {
+        ...dayofyear_year_format_options,
+        weekday: "short",
+    };
+
+    switch (type) {
+        case "time": // 01:30 PM
+            return time_format_options;
+        case "time_sec": // 01:30:42 PM
+            return {...time_format_options, second: "2-digit"};
+        case "weekday": // Wednesday
+            return weekday_format_options;
+        case "dayofyear": // Jul 27
+            return dayofyear_format_options;
+        case "dayofyear_time": // Jul 27, 01:30 PM
+            return {...dayofyear_format_options, ...time_format_options};
+        case "dayofyear_year": // Jul 27, 2016
+            return dayofyear_year_format_options;
+        case "dayofyear_year_time": // Jul 27, 2016, 01:30 PM
+            return {...dayofyear_year_format_options, ...time_format_options};
+        case "weekday_dayofyear_year": // Wednesday, July 27, 2016
+            return full_format_options;
+        case "weekday_dayofyear_year_time": // Wed, Jul 27, 2016, 13:30
+            return {...long_format_options, ...time_format_options};
+        default:
+            throw new Error("Wrong format provided.");
+    }
+}
+
+function get_user_locale(): string {
+    const user_default_language = user_settings.default_language;
+    let locale = "";
+    try {
+        locale = Intl.DateTimeFormat.supportedLocalesOf(user_default_language)[0];
+    } catch {
+        locale = "default";
+    }
```

Comment:
`empty_list[0]` does not throw an exception in JavaScript, it just gives `undefined`.

Reply:
Thanks, you're absolutely right!

I think we could write something like this instead of the snippet above:
```javascript
locale = Intl.DateTimeFormat.supportedLocalesOf(user_default_language)[0] ?? "default";
```